### PR TITLE
Add YAML Linter

### PR DESCRIPTION
**Pull request by parasense**
This was moved here from the previous PR.
The linter is based on, and adapted from, GitHub's "Simple workflow".
I didn't see a YAML linter on the GitHub actions market place portal.

Signed-off-by: Jon Disnard <jdisnard@redhat.com>

### CODE REVIEW

File: .github/workflows/lint.yaml
Comment (johnbieren) [inline .github/workflows/lint.yaml:7]: I believe the default behavior of `pull_request` is to run for all three types (`created`, `edited`, `deleted`). Not sure there is much value in running on `deleted`

File: .github/workflows/lint.yaml
Comment (johnbieren) [inline .github/workflows/lint.yaml:5]: If we already run for all PRs to main, what value do we gain from again running whenever a push happens? Unless we plan to push commits to the branch without introducing them as PRs?

File: .github/workflows/lint.yaml
Comment (johnbieren) [inline .github/workflows/lint.yaml:9]: extra white space

File: .github/workflows/lint.yaml
Comment (johnbieren) [inline .github/workflows/lint.yaml:22]: where does `GITHUB_WORKSPACE` variable come from?

**Comment by parasense** [comment]
So I wrote this on my own, it was my first GitHub action, and it's very trivial.
There are others on the GitHub actions marketplace:
https://github.com/marketplace?type=actions&query=yamllint

I would argue this one is as good as any other, and in the future I'm probably going to extend this one to feature schema validation for targeting Tekton bundle contracts:
https://tekton.dev/docs/pipelines/tekton-bundle-contracts/

So I think we should just use this, and see where it goes.

**Comment by davidmogar** [comment]
> I would argue this one is as good as any other, and in the future I'm probably going to extend this one to feature schema validation for targeting Tekton bundle contracts: https://tekton.dev/docs/pipelines/tekton-bundle-contracts/
> 
> So I think we should just use this, and see where it goes.

I disagree. We shouldn’t maintain anything that is already out there. If I’m the future we add something for Tekton schema validation, we will try to find anything existing or try to do it in a way we consume as much from Tekton as possible.

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml:4]: Apparently the default for `pull_request` events is limited to `opened`, `synchronize`, and `reopened`. 
I'm happy to make it more explicit, and any doubt or ambiguity.

Documentation link:
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml]: The idea was to allow folks to maintain a feature branch off to the side, and iterate features via successive pushes. However you make a great point, and I think this could potentially be removed. Because only our team has direct access to create branches, or directly push commits, and yet we would probably use forks anyways... so what's the point, right?

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml]: Fair enough, let's wack the white-space!

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml:16]: That `GITHUB_WORKSPACE` environment variable is setup anytime one uses GitHub's `actions/checkout@v3` default action to load the gitrepo into the runner. It's the root of the git repo inside the runner. It happens from this line:
```
steps:
      - uses: actions/checkout@v3
```

File: .github/workflows/lint.yaml
Comment (johnbieren) [reply .github/workflows/lint.yaml:4]: Oh, I missed that that was the default. Nice!

File: .github/workflows/lint.yaml
Comment (johnbieren) [reply .github/workflows/lint.yaml]: Hmm, yeah I think I see your intent and it is valid. I guess it is an opinion thing. From my perspective, if I am debugging or something and pushing code that I know isn't perfect, but will help me figure out an issue, I don't really need to be email spammed by GitHub actions telling me some static checks failed. When I am ready, because of the `workflow_dispatch` I think I can call before I open the PR to make sure I am good to go. But it also makes sense to me to just mute the emails when you are doing debug commits and to get that feedback for each commit about whether you pass static checks regardless of if your feature is ready. I am good with either, your call.

Also I am fine with you removing the `types` part of `pull_request` again. I had the default values wrong. My apologies

**Comment by parasense** [comment]
> I disagree. We shouldn’t maintain anything that is already out there. If I’m the future we add something for Tekton schema validation, we will try to find anything existing or try to do it in a way we consume as much from Tekton as possible.

Normally, I would agree 99%. However, this is the 1% where the nuances matter.
Here is what the action would look like if we used somebody else's yaml linter.
```
name: Lint
on: [push, pull_request]
jobs:
  yamllint:
    name: yamllint
    runs-on: ubuntu-latest
    steps:
      - name: ⤵️ Check out code from GitHub
        uses: actions/checkout@v2
      - name: 🚀 Run yamllint
        uses: frenck/action-yamllint@v1
```
That's 11 lines of action script, taken from here: https://github.com/frenck/action-yamllint

However, our linter currently looks like this:
```
name: Linters
on:  # yamllint disable-line rule:truthy
  pull_request:
    branches: ['main']
    types: ['opened', 'reopened', 'synchronize']
  workflow_dispatch:
jobs:
  yamllint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install yamllint
        run: sudo apt-get install yamllint
      - name: Run yamllint
        run: yamllint --format github $GITHUB_WORKSPACE
```

It's effectively the same amount of lines and effort. By using somebody else's linter we would gain the advantage of subtracting about ~two lines, and in my humble opinion that is not very persuasive. However, I think your point is valid about maintaining code, less is more... and if this action does grow it makes sense to consolodate.

So after some consideration, I've decided to maintain this linter. I'll post it to the github actions market place using either in my personal profile, or (hopefully) hosted on Red Hat's action repo: https://github.com/orgs/redhat-actions/repositories?q=&type=all

For the time being I think it's reasonable to move forward.

**Comment by davidmogar** [comment]
> Normally, I would agree 99%. However, this is the 1% where the nuances matter. Here is what the action would look like if we used somebody else's yaml linter.
> 
> It's effectively the same amount of lines and effort. By using somebody else's linter we would gain the advantage of subtracting about ~two lines, and in my humble opinion that is not very persuasive. However, I think your point is valid about maintaining code, less is more... and if this action does grow it makes sense to consolidate.

I still disagree.

You would also gain the advantage of not having to maintain anything. Let's say that the package name changes in the future or there's a better package for this. The maintainers of the action will take care of it and you will have to do nothing. And they will probably do that in a way that nothing breaks for you as GitHub actions maintain at least two ubuntu image versions, so they can work in advance.

On top of that, the [existing action](https://github.com/marketplace/actions/yaml-lint#optional-input-parameters) already has functionality that yours doesn't have. That is strict mode, config files, formatting...

I don't see a need to reinvent the wheel.

Regarding Tekton resources validation, that is not that simple. Usually you would use something like [kubeval](https://github.com/instrumenta/kubeval). The problem with this approach is that it doesn't validate CRD, although I'm quite sure something like [this](https://github.com/instrumenta/kubeval/issues/47#issuecomment-1032826047) would allow us to validate CRDs. Now, the real problem is that Tekton CRDs don't seem to have the whole spec schema, so validations would always work.

So, let's put this aside and focus on pushing the bundles.

File: .github/workflows/lint.yaml
Comment (davidmogar) [inline .github/workflows/lint.yaml:13]: This is not needed as the yamllint actions already installs yamllint.

File: .github/workflows/tekton_bundle_push.yaml
Comment (davidmogar) [inline .github/workflows/tekton_bundle_push.yaml:84]: Not strictly true. You can pass a file with all the resources separated by `---` and all of them will get processed. It works exactly in the same way than kubectl/oc.

File: .github/workflows/tekton_bundle_push.yaml
Comment (davidmogar) [inline .github/workflows/tekton_bundle_push.yaml:1]: I wouldn't expect to see this file in a PR for adding a yaml linter. This should be in a different PR.

File: definitions/hello-release/hello-release.yaml
Comment (davidmogar) [inline definitions/hello-release/hello-release.yaml:1]: I wouldn't expect to see this file in a PR for adding a yaml linter. This should be in a different PR or completely removed as we don't really need this anymore.

File: definitions/hello-world/hello-world.yaml
Comment (davidmogar) [inline definitions/hello-world/hello-world.yaml:1]: I wouldn't expect to see this file in a PR for adding a yaml linter. This should be in a different PR or completely removed as we don't really need this anymore.

**Review by davidmogar — CHANGES_REQUESTED** [review]

File: .github/workflows/lint.yaml
Comment (johnbieren) [reply .github/workflows/lint.yaml:7]: I think it is nice to leave this in here. It doesn't hurt anything and allows others to use it if they want to (like I wanted to use application-service workflows earlier on)

File: .github/workflows/tekton_bundle_push.yaml
Comment (johnbieren) [reply .github/workflows/tekton_bundle_push.yaml:1]: On a related note, this PR has 20 commits. Is that expected? Maybe it should be squashed down a bit?

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml:13]: You're correct, and actually it seems `yamllint` is installed by default in the main images used in github actions, so even without the 3rd party action it would have worked (supposedly).

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml:7]: I like being able to run the workflow manually, say for example in a feature branch outside the normal PR workflow.

File: .github/workflows/tekton_bundle_push.yaml
Comment (parasense) [reply .github/workflows/tekton_bundle_push.yaml:1]: > I wouldn't expect to see this file in a PR for adding a yaml linter. This should be in a different PR.

The PR adds configuration for where to find the ".yamllint" config file, so in my mind it's fair game to also include a config. Also I don't think it's fair to say the inclusion of the config file is somehow overloading this PR.

File: .github/workflows/tekton_bundle_push.yaml
Comment (parasense) [reply .github/workflows/tekton_bundle_push.yaml:84]: To clarify, 
What does NOT work is this hypothetical example, where you may assume there are multiple YAML files:
```
tkn bundle push  -f ./definition/hello-world/*.yaml  "hello-world:main"
```

File: definitions/hello-release/hello-release.yaml
Comment (parasense) [reply definitions/hello-release/hello-release.yaml:1]: You can safely ignore this file, it's from the main branch, pulled in via a rebase.

This PR was rebased on main.
That is why you see superfluous commits from the main branch.
What I do is add multiple remotes in my local repo:

My fork remote looks like this:
```
$ git remote get-url origin
git@github.com:parasense/release-bundles.git
```

So I just add another remote for the upstream of my fork, and then rebase from that.
```
git remote add upstream https://github.com/release-service/release-bundles
git fetch upstream
git rebase upstream/main
git push origin/branch-name
```

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml]: fixed

File: .github/workflows/lint.yaml
Comment (parasense) [reply .github/workflows/lint.yaml]: Fair enough....

However, I have a counter-point: I think line-breaks improve readability. Consider that YAML files are not purely for machine reading, which is why the yaml linter has rules for line length, and for N (consecutive) line-breaks are allowable, etc.. There is no imperative reason to NOT have line breaks, or long lines; nor is there any compelling argument for the reverse.  So please, I beg your pardon  but I consider this to be needlessly pedantic. 

That said, I'm happy to remove the line-break, as I'm utterly indifferent. ;-)
(I believe as of the latest commit, there shouldn't be any)

File: .github/workflows/lint.yaml
Comment (davidmogar) [reply .github/workflows/lint.yaml:7]: That's sooooo odd. It says that I made that comment yesterday but I didn't. I remember making the comment looong ago. Anyway, good to me.

File: .github/workflows/lint.yaml
Comment (davidmogar) [reply .github/workflows/lint.yaml]: You have the dashes for that. Let's stick to the style in the official docs.

File: .github/workflows/tekton_bundle_push.yaml
Comment (davidmogar) [reply .github/workflows/tekton_bundle_push.yaml:1]: My comment is not in the config file. It is in the Tekton Bundle Push workflow which should not be part of a PR to add yaml linting.

Regarding the rebase, those commits shouldn't be here if they were already in main. Please, clean up the PR and squash the commits so we can review properly.

File: definitions/hello-release/hello-release.yaml
Comment (davidmogar) [reply definitions/hello-release/hello-release.yaml:1]: Rebase doesn't work like that. Please, fix this PR so it only contains the relevant stuff.

File: definitions/hello-world/hello-world.yaml
Comment (davidmogar) [reply definitions/hello-world/hello-world.yaml:1]: Same comment above.

**Review by davidmogar — APPROVED** [review]
Looking good to me. Please, don't merge this without squashing.

**Comment by johnbieren** [comment]
I don't see anything requesting change. I have the ability to merge and it is green. But, as David said, we don't want to merge this as 22 commits. They need to be squashed

**Pull request merged**
